### PR TITLE
SI-7788 Predef.conforms can circumvent implicit shadowing

### PR DESCRIPTION
**Pull request by retronym**
... together with small improvements to shadowing detection
in implicit typechecking.

Review by @adriaanm

### CODE REVIEW

**Comment by adriaanm** [comment]
I think this needs to be rebased -- due to the askLoadedType change

**Comment by retronym** [comment]
Rebased

**Comment by adriaanm** [comment]
Darn, and now it needs a check file update

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -591,24 +593,26 @@ trait Implicits {
...
-        } else {
-          assert(info.pre != NoPrefix, info)
-          // SI-2405 Not info.name, which might be an aliased import
-          val implicitMemberName = info.sym.name
-          Select(gen.mkAttributedQualifier(info.pre), implicitMemberName)
-        }
-      }
-      val itree1 = if (isBlackbox(info.sym)) suppressMacroExpansion(itree0) else itree0
-      typingLog("considering", typeDebug.ptTree(itree1))
+      val itree0 = atPos(pos.focus)(gen.mkAttributedRef(info.pre, info.sym, info.name))
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: why not info.sym.name? or better yet, why use the name at all if we have the symbol?

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: I tried out the test case from SI-2405 and it seems to work, but the :implicits command doesn't see the renamed implicit (it doesn't see it in 2.11.0-SNAPSHOT either, so never mind). When importing without renaming, it works.

```
$ cat ~/.sbt/0.13/pr-validation-resolver.sbt
resolvers += "pr-scala" at "http://private-repo.typesafe.com/typesafe/scala-pr-validation-snapshots/"
$ sbt
> set scalaVersion := "2.11.0-1f942ed-SNAPSHOT"
> console
Welcome to Scala version 2.11.0-1f942ed-SNAPSHOT (Java HotSpot(TM) 64-Bit Server VM, Java 1.6.0_65).
Type in expressions to have them evaluated.
Type :help for more information.

scala> object A { implicit val x = 1 }
defined object A

scala> import A.{x => y}
import A.{x=>y}

scala> implicitly[Int]
res0: Int = 1

scala> :implicits
No implicits have been imported other than those in Predef.

scala> import A.x

scala> :implicits
/* 1 implicit members imported from A */
  /* 1 defined in A */
  implicit val x: Int
```

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: > why not info.sym.name? or better yet, why use the name at all if we have the symbol?

It makes the code more robust against `resetLocalAttrs` and a retypecheck, which is prone to happen in macros.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: Righto. Does the invariant `info.sym.name == info.name` hold?

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala]: Nope. `info.name` is the potentially-aliased name by which the implicit is in scope; `info.sym.name` is the declared name.

**Comment by retronym** [comment]
I've rebased with the checkfile fixup, and added a comment about renamed imports.

**Comment by adriaanm** [comment]
LGTM -- :star: for the comment!

**Comment by adriaanm** [comment]
Rebuild when https://github.com/scala-ide/uber-build/pull/26 is merged.

**Comment by adriaanm** [comment]
I looked into what could be causing the failure from our side, but it's failing across all of our jenkins lampscalatst\* slaves, so not sure what to do there. It's not like they're running out of disk space. My hypothesis is that the build script is not robust against aborts.

**Comment by adriaanm** [comment]
PLS REBUILD ALL

**Comment by dragos** [comment]
The failures seem legit: refactoring is failing tests in organize imports, related to implicits:

```
OrganizeImportsFullyRecomputeTest.importMethodFromSamePackage:599 expected:<...e a.b.c

    import []TestImplicits.string...> but was:<...e a.b.c

    import [a.b.c.]TestImplicits.string...>
```

Not sure if the changes are intentional, and scala-refactoring should update its test, or a regression. /cc @misto

**Comment by misto** [comment]
Hard to tell, but I think it could be my test that's at fault. Do you want me to adapt it to the new behaviour?

**Comment by adriaanm** [comment]
This is a new failure -- we didn't get there last time.

**Comment by retronym** [comment]
NOLITTER!
PLS REBUILD ALL

**Comment by retronym** [comment]
The breakage in scala-refactoring comes down to different, though semantically identical, trees:

``` scala
    package a.b.c
    object TestImplicits {
      implicit def stringToBytes(s: String): Array[Byte] = s.getBytes
    }
```

``` scala
// Old
TestImplicits.stringToBytes("")

// New
c.this.TestImplicits.stringToBytes("")
```

I'll change this PR to avoid the gratuitous change.

**Comment by retronym** [comment]
closing for a quick reno...

**Comment by retronym** [comment]
```
% git log --oneline master...
6660f70  Add a clarifying comment
63ddc3d  Refactoring in typedImplicit1
aa6da02  Predef.conforms can circumvent implicit shadowing
7838b44 Improve shadowed implicits detection, reporting.
* 2355f0d Elide Select(package, ...) in TreeGen
* ed11d71  Pin down the precise shape synthetic implicit tree
```

@adriaanm The first two commits, starred above, are new in the latest `push -f`.

**Comment by dragos** [comment]
Looks like `Predef.String` is added by _organize imports_ refactoring. @misto, could you please have a look?

**Comment by retronym** [comment]
I believe that my changes conflict with a bit of ad-hockery in scala-refactoring.

``` scala
  def isQualifierNotDefaultImported(t: Tree) = t match {
    case t: Select =>
      val Scala = newTypeName("scala")
      val Java = newTypeName("java")
      val `lang` = newTermName("lang")
      t.qualifier match {
        case Ident(Names.scala) => false
        case This(Scala) => false
        case Select(This(Java), `lang`) => false
        case Select(Ident(Names.scala), Names.pkg) => false
        case Select(Ident(Names.scala), Names.Predef) => false
        case Select(This(Scala), _) => false
        case qual if qual.symbol.isSynthetic && !qual.symbol.isModule => false
        case qual if qual.nameString == "ClassTag" => false
        case _ => true
      }
    case _ => true
  }

```

**Comment by misto** [comment]
Oh yes... does the compiler have a handy method to determine if `t.qualifier` is from Predef or any other default-imported place?

**Comment by retronym** [comment]
@misto What decisions are based on this function? 

We've got:

``` scala
    /** Conditions where we omit the prefix when printing a symbol, to avoid
     *  unpleasantries like Predef.String, $iw.$iw.Foo and <empty>.Bippy.
     */
    final def isOmittablePrefix = /*!settings.debug.value &&*/ (
         UnqualifiedOwners(skipPackageObject)
      || isEmptyPrefix
    )

    // Modules whose members are in the default namespace
    // SI-5941: ScalaPackage and JavaLangPackage are never ever shared between mirrors
    // as a result, `Int` becomes `scala.Int` and `String` becomes `java.lang.String`
    // I could just change `isOmittablePrefix`, but there's more to it, so I'm leaving this as a todo for now
    lazy val UnqualifiedModules = List(PredefModule, ScalaPackage, JavaLangPackage)
    // Those modules and their module classes
    lazy val UnqualifiedOwners  = UnqualifiedModules.toSet ++ UnqualifiedModules.map(_.moduleClass)
```

**Comment by misto** [comment]
That's sounds promising, I'll give it a try!

**Comment by misto** [comment]
It almost works! There seems to be only one case in my tests where it doesn't.

`class Blah {  val x = for (e <- Array(1, 2, 3)) yield e }`

Here isOmittablePrefix on the qualifier of `scala.this.Array.canBuildFrom` returns false.

**Comment by retronym** [comment]
@misto Could you demonstrate that in a `:power` mode session?

```
scala> typeOf[scala.Array.type].prefix.typeSymbol.thisType
res35: $r.intp.global.Type = scala.type

scala> typeOf[scala.Array.type].prefix
res36: $r.intp.global.Type = scala.type

scala> typeOf[scala.Array.type].prefix.typeSymbol.isOmittablePrefix
res37: Boolean = true

scala> showRaw(gen.mkAttributedRef(typeOf[scala.Array.type].member(newTermName("canBuildFrom"))), printIds = true, printTypes = true)
res38: String =
Select[1](This[2](TypeName("Array")), TermName("canBuildFrom")#23476)
[1] PolyType(List(TypeName("T")#23478), MethodType(List(TermName("t")#23479), TypeRef(ThisType(scala.collection.generic#5751), scala.collection.generic.CanBuildFrom#21152, List(ExistentialType(List(TypeName("_$4")#23477), TypeRef(ThisType(scala#21), scala.Array#690, List(TypeRef(NoPrefix, TypeName("_$4")#23477, List())))), TypeRef(NoPrefix, TypeName("T")#23478, List()), TypeRef(ThisType(scala#21), scala.Array#690, List(TypeRef(NoPrefix, TypeName("T")#23478, List())))))))
[2] ThisType(scala.Array#692)
```

**Comment by misto** [comment]
I just noticed that I tested it on 2.10, should it make a difference?

```
scala> val cbf = gen.mkAttributedRef(typeOf[scala.Array.type].member(newTermName("canBuildFrom"))).asInstanceOf[Select]
cbf: $r.intp.global.Select = Array.this.canBuildFrom

scala> cbf.qualifier
res10: $r.intp.global.Tree = Array.this

scala> cbf.qualifier.symbol
res11: $r.intp.global.Symbol = object Array

scala> cbf.qualifier.symbol.isOmittablePrefix
res12: Boolean = false
```

Edit: It's the same with 2.11.0-M7

**Comment by retronym** [comment]
`Array` isn't omittable as a prefix, on `scala` and `Predef`.

**Comment by misto** [comment]
Ah, no of course not, I'm an idiot :-) Well then I can just use `isOmittablePrefix` and save me the trouble.

**Comment by retronym** [comment]
@misto Does that change make scala-refactoring compatible before/after this PR?

BTW, it's easy to test this with:

```
% git diff head
diff --git a/pom.xml b/pom.xml
index b157e23..a03caa7 100644
--- a/pom.xml
+++ b/pom.xml
@@ -237,6 +237,11 @@
       <name>Sonatype maven snapshot repository</name>
       <url>https://oss.sonatype.org/content/repositories/snapshots</url>
     </repository>
+    <repository>
+      <id>pr.snapshot</id>
+      <name>PR snapshot repository</name>
+      <url>http://private-repo.typesafe.com/typesafe/scala-pr-validation-snapshots/</url>
+    </repository>
   </repositories>
   <pluginRepositories>
     <pluginRepository>

% mvn -Dscala.version=2.11.0-239cc1f-SNAPSHOT test
```

**Comment by misto** [comment]
Yes it does! There are test failures with 239cc1f but it's fine with ed11d71!

**Comment by misto** [comment]
I've pushed the changes to scala-refactoring.

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
NOLITTER!

**Comment by retronym** [comment]
I hope that my flailing around in this PR is of some comedic value. I can clearly see what appears to be a bona-fide dawn now, so stay tuned for more.

**Comment by xeno-by** [comment]
@retronym Good luck! Keeping fingers crossed.

**Comment by xeno-by** [comment]
@retronym Congrats!

**Comment by retronym** [comment]
@xeno-by Thanks! Perhaps you could help out to review, in particular that the last two commits?

**Comment by xeno-by** [comment]
How about a different name for `itree0`?

**Comment by retronym** [comment]
Shutting this down again thanks to the diligent reviewer poking a hole in the core change...

**Comment by adriaanm** [comment]
@retronym, Sorry, I haven't re-read the whole thread, but how about deprecating Predef.conforms, removing its implicit status, and adding a `implicit def __$MAGIC$__conforms`  to Predef?

**Pull request closed without merging**